design/one-pager-strongly-typed-class.md

```diff
@@ -0,0 +1,156 @@
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+  labels:
+    mysqlinstance.storage.crossplane.io/default: "true"
+parameters:
+  class: db.t2.small
+  masterUsername: masteruser
+  securityGroups: "sg-ab1cdefg,sg-05adsfkaj1ksdjak"
+  size: "20"
+provisioner: rdsinstance.database.aws.crossplane.io/v1alpha1
+providerRef:
+  name: aws-provider
+reclaimPolicy: Delete
+```
+
+Resource Claim:
+
+```yaml
+apiVersion: storage.crossplane.io/v1alpha1
+kind: MySQLInstance
+metadata:
+  name: mysql-claim
+  namespace: demo
+spec:
+  engineVersion: "9.6"
+```
+
+This model is powerful because it allows an application developer to create a resource claim without having to know the implementation details or even the underlying provider. However, the fact that every resource class if of the same `kind` presents a key issue: The required parameters for a resource class may vary widely, and they are currently only provided as an arbitrary map that is eventually read by the controller for the the specified `provisioner`. Therefore, an administrator who is creating resource classes does not know what fields are required and will not be notified of missing or extraneous fields until the provisioning of a resource that references the class.
+
+This problem can be solved by the implementation of "strongly typed" resource classes. Instead of every resource class being of `kind: ResourceClass`, they would map one-to-one with the concrete managed resource kind for which they provided specifications (i.e. `kind: RDSInstanceClass`).
+
+Unfortunately, this makes it much more difficult to identify default resource classes because the default class controller currently watches for objects of `kind: ResourceClass`. With an unknown number of possible kinds of resource classes that could be installed and serve as default for a claim `kind`, a default class controller cannot determine all class kinds to watch.
+
+One solution would instead be to map resource classes one-to-one with claim kinds (i.e. `MySQLInstanceClass`), but that results in the same issue of varying required parameters that is currently being experienced.
+
+## Proposal
+
+In order to satisfy both the requirements of strong typing and default resource classes, a new custom resource kind, `Defaulter`, can be implemented to specify when a specific resource class kind can serve as default for a given claim kind. The workflow would proceed as follows:
+
+1. An admin installs Crossplane, which specifies the portable claim kinds that are available as part of the core Crossplane.
+2. An admin installs an infrastructure stack (i.e. a set of CRDs and controllers for a provider's managed service offerings).
+3. An admin creates an instance of an installed resource class CRD (i.e. `kind: RDSInstanceClass`). It could look as follows:
+
+```yaml
+apiVersion: database.aws.crossplane.io/v1alpha1
+kind: RDSInstanceClass
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+class: db.t2.small
```

Comment:
I had that same internal debate. I would prefer to move it to a `spec` because it is more intuitive to me, but I agree that it isn't exactly what `spec` usually means.